### DISCUSSION TITLE
How to Compute Directional Derivatives or Derivatives with Respect to Coordinates x and y of a Field in MOOSE ?

### DISCUSSION THREAD

**Discussion by mahdipasaie** [idaholab/moose#28637] [Q&A Modules: Phase field] [answered]
Hi, 

I am working on implementing a phase field equation in MOOSE that involves directional derivatives of a field $\phi$. Specifically, I need to compute the derivatives of  $\phi$  with respect to the \( x \)- and \( y \)-coordinates (i.e., $\frac{\partial \phi}{\partial x}$ and $\frac{\partial \phi}{\partial y}$).

These derivatives are required to calculate an angular-dependent function, $\theta$, defined as:

**$$
\theta = \arctan \left( \frac{\frac{\partial \phi}{\partial y}}{\frac{\partial \phi}{\partial x}} \right)
$$**

I need to define $\theta$ as a material property in MOOSE, but I'm not sure how to compute the directional derivatives in this framework. Could someone guide me on how to obtain these derivatives and use them in the material property definition?

Here is the full equation if anyone is interested:

$$
\tau_0 a(\theta)^2 \frac{\partial \phi}{\partial t} = W_0^2 \nabla \cdot \left[a(\theta)^2 \nabla \phi\right] - \frac{\partial}{\partial x}\left[a(\theta) a^{\prime}(\theta) \frac{\partial \phi}{\partial y}\right] + \frac{\partial}{\partial y}\left[a(\theta) a^{\prime}(\theta) \frac{\partial \phi}{\partial x}\right] + \left[\phi - \lambda u(1 - \phi^2)\right]\left(1 - \phi^2\right)
$$

where:
- $\ a(\theta) = 1 + \epsilon \cos(4 \theta) \$
- $\ \theta = \arctan\left(\frac{\frac{\partial \phi}{\partial y}}{\frac{\partial \phi}{\partial x}}\right) \$

The coupled equation:

$$
\frac{\partial u}{\partial t} = D \nabla^2 u + \frac{1}{2} \frac{\partial h(\psi)}{\partial t}
$$

Any examples or pointers would be greatly appreciated. Thanks in advance!

For reference, the equations I'm working with are Equations 87 and 88 from *"Quantitative Phase-field Modeling of Dendritic Growth in Two and Three Dimensions"* by Alain Karma. You can find the article [here](https://journals.aps.org/pre/abstract/10.1103/PhysRevE.57.4323)

Mahdi.

**Comment by GiudGiud** [comment]
Hello

Directional derivatives are computed using a chain rule.
so
```
da/dtheta = da/dx dx/dtheta + da/dy dy/theta
```
```
dx/dtheta = 1 / dtheta/dx
```
here

**Reply by mahdipasaie** [reply]
thank you for your reply. To define theta, is it possible that I extract the x and y components of the phi as ```_grad_u [_qp](0)``` and ```_grad_u[_qp](1)``` then define these as material properties and use them to define theta? 

and my second question is for the terms $\frac{\partial}{\partial x}\left[a(\theta) a^{\prime}(\theta) \frac{\partial \phi}{\partial y}\right]$  and $\frac{\partial}{\partial y}\left[a(\theta) a^{\prime}(\theta) \frac{\partial \phi}{\partial x}\right]$ : 

$a(\theta) a^{\prime}(\theta) \frac{\partial \phi}{\partial x}\ = term1$

$a(\theta) a^{\prime}(\theta) \frac{\partial \phi}{\partial y}\ = term2$


when I get the weak forms of these terms the partial derivative would go on the test function I mean the weak form is: 

$< term1 , \frac{\partial \ Testfunction}{\partial y}\ >$  

and 

$< term2 , \frac{\partial \ Testfunction}{\partial x}\ >$

so is it possible to multiply ( inner product )  the x and y components of the test function in moose? is it ``` _grad_test[_i][_qp](0)``` and  ```_grad_test[_i][_qp](1)```. 

Thanks again.

**Reply by GiudGiud** [reply]
if u is a variable,
you should be able to compute grad_u directly in any material property using the coupledGradient routine from the Coupleable API, no need to save them as material properties

> so is it possible to multiply ( inner product ) the x and y components of the test function in moose?

yes. You can index into grad_test[_i][_qp] like you did

**Comment by mahdipasaie** [comment, marked as answer]
So this is how I define Theta in FEM: 

- $\ a(\theta) = 1 + \epsilon \cos(4 \theta) \$

```cpp
#include "ADtheta.h"
#include "MooseMesh.h"
#include "MathUtils.h"

registerMooseObject("diffApp", ADtheta);

InputParameters
ADtheta::validParams()
{
    InputParameters params = Material::validParams();
    params.addClassDescription("2D interfacial anisotropy");
    params.addParam<Real>(
        "epsilon", 0.02, "Strength of the anisotropy");
    params.addParam<Real>(
        "w0", 1.0, "Interface width");
    params.addRequiredCoupledVar("op", "Order parameter defining the solid phase");
    
  return params;
}

ADtheta::ADtheta(const InputParameters & parameters)
  : Material(parameters),
    _epsilon(getParam<Real>("epsilon")),
    _w0(getParam<Real>("w0")),
    _eps(declareADProperty<Real>("eps")),
    _deps(declareADProperty<Real>("deps")),
    _op(adCoupledValue("op")),
    _grad_op(adCoupledGradient("op"))
{
  // this currently only works in 2D simulations
  if (_mesh.dimension() != 2)
    mooseError("ADInterfaceOrientationMaterial requires a two-dimensional mesh.");
}


void
ADtheta::computeQpProperties()
{
    const ADReal grad_x_phi = _grad_op[_qp](0);  // ∂x φ
    const ADReal grad_y_phi = _grad_op[_qp](1);  // ∂y φ

    // Compute the magnitude of the gradient
    const ADReal grad_norm = std::sqrt(grad_x_phi * grad_x_phi + grad_y_phi * grad_y_phi);

    // Initialize theta
    ADReal theta = 0.0;

    // Avoid undefined atan2 when gradient is zero
    if (grad_norm > 1e-14) // Use a small tolerance to avoid division by zero
    {
        theta = std::atan2(grad_y_phi, grad_x_phi);
    }

    // Compute anisotropy functions
    _eps[_qp] = _w0 * (1.0 + _epsilon * std::cos(4.0 * theta));
    _deps[_qp] = _w0 * (- _epsilon * 4.0 * std::sin(4.0 * theta));
}
```

and in the input file: 

[Materials]
    [./Theta]
        type = ADtheta
        op = phi
    [../]
[]

**Reply by GiudGiud** [reply]
what is deps again? deps/d_theta?

for this line
```
   if (grad_norm > 1e-14) // Use a small tolerance to avoid division by zero
    {
        theta = std::atan2(grad_y_phi, grad_x_phi);
```
just make sure you are ok with the limit behaviors listed here
https://en.cppreference.com/w/cpp/numeric/math/atan2

**Reply by mahdipasaie** [reply]
Yes, `deps` represents the derivative of $a(\theta)$ with respect to $\theta$. I should also mention that I reviewed the pahse filed source code on MOOSE's GitHub, which gave me some insights. However, I found it overly complex for my specific needs, so I opted to write my own implementation. That said, there might still be some details that I didn't fully grasp.

Regarding the tolerance check, I initially expected the code to work without the if statement. However, without it, the simulation wasn't converging. That said, I made several parameter changes afterward to get the simulation working, so I’ll need to revisit this check to see if it's still necessary under the updated configuration

<img width="826" alt="Screenshot 2024-09-30 at 1 05 12 PM" src="https://github.com/user-attachments/assets/285b2bd2-4ab2-4296-bd29-79723504960d">
<img width="826" alt="Screenshot 2024-09-30 at 1 05 30 PM" src="https://github.com/user-attachments/assets/156893c3-3e59-4d7a-b9d9-52e86b27487f">

**Reply by mahdipasaie** [reply]
I'd like to also test the Finite Volume Method (FVM) kernel in MOOSE by implementing the same physics, but this time using FVM. Could you help me with defining $a(\theta)$ in an FVM kernel, similar to how it's defined in the current setup?

**Reply by GiudGiud** [reply]
you can use the same material properties in finite volume simulations
The kernels are a little different, as in FV you implement the strong form rather than the weak form in the kernels
see
https://mooseframework.inl.gov/moose/finite_volumes/fv_design.html
https://mooseframework.inl.gov/moose/syntax/FVKernels/

**Reply by mahdipasaie** [reply]
It seems that $a(\theta)$ should now be a `FunctorMaterial` in order to be used within a kernel for a term like the one shown below in the Finite Volume Method (FVM):

$$
\nabla \cdot \left[a(\theta)^2 \nabla \phi\right]
$$

This term in my PDE behaves similarly to the `FVDiffusion.C` kernel in MOOSE, where $a(\theta)$ plays a role analogous to the `coeff` in that kernel. In `FVDiffusion.C`, the coefficient is interpolated to the faces of the elements. This interpolation process can be handled if `Coeff` is a `FunctorMaterial` object in MOOSE, correct?

Here’s the relevant part of the kernel implementation for reference:
```cpp 

ADReal
FVDiffusion::computeQpResidual()
{
  using namespace Moose::FV;
  const auto state = determineState();

  auto dudn = gradUDotNormal(state);
  ADReal coeff;

  // If we are on internal faces, we interpolate the diffusivity as usual
  if (_var.isInternalFace(*_face_info))
  {
    const ADReal coeff_elem = _coeff(elemArg(), state);
    const ADReal coeff_neighbor = _coeff(neighborArg(), state);
    // If the diffusion coefficients are zero, then we can early return 0 (and avoid warnings if we
    // have a harmonic interpolation)
    if (!coeff_elem.value() && !coeff_neighbor.value())
      return 0;

    interpolate(_coeff_interp_method, coeff, coeff_elem, coeff_neighbor, *_face_info, true);
  }
  // Else we just use the boundary values (which depend on how the diffusion
  // coefficient is constructed)
  else
  {
    const auto face = singleSidedFaceArg();
    coeff = _coeff(face, state);
  }

  return -1 * coeff * dudn;
}
```

**Reply by GiudGiud** [reply]
some of the FVKernels do use functor materials. there are still a few using regular material properties. It's as you want
